middle_end/flambda2.0/simplify/simplify_binary_primitive.ml

```diff
@@ -845,9 +851,29 @@ let simplify_immutable_block_load ~result_kind dacc ~original_term _dbg
     let n =
       Targetint.OCaml.add (Target_imm.to_targetint index) Targetint.OCaml.one
     in
+    (* CR mshinwell: We should be able to use the size in the [access_kind]
+       to constrain the type of the block *)
+    let tag : _ Or_unknown.t =
+      match access_kind with
+      | Values { tag; _ } ->
+        (* We are dealing here with blocks, not arrays, so even if the size
+           of the block is zero the tag is still as expected. *)
+        Known (Tag.Scannable.to_tag tag)
+      | Naked_floats { size; } ->
+        (* Arrays of naked floats, whether or not the float array
+           optimisation is in effect, always have tag zero.  This means that
+           if we don't know the size here, we don't know the tag. *)
```

Comment:
This comment confused me, until I understood that it's only zero-sized float arrays that have tag zero.

Reply:
I'm going to put the `Naked_floats` case first, which will mean that the associated comment comes first, noting this peculiarity.

Reply:
Actually, this doesn't address my question: if I understand correctly, we're in the regular block case, and blocks of naked floats have tag double_array_tag, unless they are of size 0, in which case they have tag 0. This seems to be what the code is doing, but the comment is just confusing.